src/lib/device/journal_vdev.cpp

```diff
@@ -34,75 +34,211 @@
+        auto next_chunk = data->next_chunk;
 
-#ifndef NDEBUG
-    if (end_offset < ds_off) { HS_DBG_ASSERT_EQ(size() - used_size(), static_cast< uint64_t >(ds_off - end_offset)); }
-#endif
-    // works for both "end_offset >= ds_off" and "end_offset < ds_off";
-    if (offset_in_chunk + sz <= chunk->size()) {
-        // not acrossing boundary, nothing to do;
-    } else if ((used_size() + (chunk->size() - offset_in_chunk) + sz) <= size()) {
-        // across chunk boundary, still enough space;
+        // Clear the private chunk data.
+        *data = JournalChunkPrivate{};
+        update_chunk_private(chunk, data);
+
+        // Try add to chunk_pool, if pool is full we remove the chunk.
+        if (m_chunk_pool->size() < m_chunk_pool->capacity()) {
+            m_chunk_pool->enqueue(chunk);
+            LOGWARN("Reuse to pool orphan chunk {} found for logdev {} next {}.", chunk_id, logdev_id, next_chunk);
+        } else {
+            LOGWARN("Removing orphan chunk {} found for logdev {} next {}.", chunk_id, logdev_id, next_chunk);
+            m_dmgr.remove_chunk(chunk);
+        }
+    }
 
-        // Update the overhead to total write size;
-        m_write_sz_in_total.fetch_add(chunk->size() - offset_in_chunk, std::memory_order_relaxed);
+    // Start the chunk pool.
+    m_chunk_pool->start();
+}
 
-        // If across chunk boundary, update the chunk super-block of the chunk size
-        chunk->update_end_of_chunk(offset_in_chunk);
+void JournalVirtualDev::update_chunk_private(shared< Chunk >& chunk, JournalChunkPrivate* private_data) {
+    sisl::blob blob;
+    blob.bytes = r_cast< uint8_t* >(private_data);
+    blob.size = sizeof(JournalChunkPrivate);
+    chunk->set_user_private(blob);
+}
 
-#ifdef _PRERELEASE
-        iomgr_flip::test_and_abort("abort_after_update_eof_cur_chunk");
-#endif
-        // get next chunk handle
-        auto next_chunk = get_next_chunk(chunk);
-        if (next_chunk != chunk) {
-            // Since we are re-using a new chunk, update this chunk's end as its original size;
-            next_chunk->update_end_of_chunk(chunk->size());
+uint64_t JournalVirtualDev::get_end_of_chunk(shared< Chunk >& chunk) const {
+    auto* private_data = r_cast< JournalChunkPrivate* >(const_cast< uint8_t* >(chunk->user_private()));
+    return private_data->end_of_chunk_size;
+}
+
+shared< JournalVirtualDev::Descriptor > JournalVirtualDev::open(logdev_id_t logdev_id) {
+    auto it = m_journal_descriptors.find(logdev_id);
+    if (it == m_journal_descriptors.end()) {
+        auto journal_desc = std::make_shared< JournalVirtualDev::Descriptor >(*this, logdev_id);
+        m_journal_descriptors.emplace(logdev_id, journal_desc);
+        return journal_desc;
+    }
+
+    HS_LOG(TRACE, device, "Opened log device descriptor {}", logdev_id);
+    return it->second;
+}
+
+void JournalVirtualDev::Descriptor::append_chunk() {
+    // Get a new chunk from the pool.
+    auto new_chunk = m_vdev.m_chunk_pool->dequeue();
```

Comment:
will there be a case when m_chunk_pool is empty and new_chunk is null?

Reply:
Added assert.